orm/model_bucket.go

```diff
@@ -0,0 +1,174 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// TODO
+// - migrations
+// - do not use Bucket but directly access KVStore
+// - register for queries
+
+// Model is impelemented by any entity that can be stored using ModelBucket.
+//
+// This is the same interface as CloneableData. Using the right type names
+// provides an easier to read API.
+type Model interface {
+	weave.Persistent
+	Validate() error
+	Copy() CloneableData
+}
+
+// ModelSlicePtr represents a pointer to a slice of models. Think of it as
+// *[]Model Because of Go type system, using []Model would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type ModelSlicePtr interface{}
+
+// ModelBucket is implemented by buckets that operates on Models rather than
+// Objects.
+type ModelBucket interface {
+	// One query the database for a single model instance. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// model.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given model type cannot be used to contain stored entity, ErrType
+	// is returned.
+	One(db weave.ReadOnlyKVStore, key []byte, dest Model) error
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is retured and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest ModelSlicePtr) error
+
+	// Put saves given model in the database. Before inserting into
+	// database, model is validated using its Validate method.
+	// If the key is nil or zero length then a sequence generator is used
+	// to create a unique key value.
+	// Using a key that alreayd exists in the database cause the value to
+	// be overwritten.
+	Put(db weave.KVStore, key []byte, m Model) ([]byte, error)
+
+	// Delete removes an entity with given primary key from the database.
+	// It returns ErrNotFound if an entity with given key does not exist.
+	Delete(db weave.KVStore, key []byte) error
+
+	// Register registers this buckets content to be accessable via query
+	// requests under the given name.
+	Register(name string, r weave.QueryRouter)
+}
+
+// NewModelBucket returns a ModelBucket instance. This implementation relies on
+// a bucket instance. Final implementation should operate directly on the
+// KVStore instead.
+func NewModelBucket(b Bucket) ModelBucket {
+	return &modelBucket{
+		b:     b,
+		idSeq: b.Sequence("id"),
```

Comment:
So this automatically is an IdGen model?

Reply:
Yes.
When using `Bucket` you always inherit and optionally attach the sequence to your implementation if you need it. This makes it clean and optional.
I was wondering how to do the same for the `ModelBucket`, without introducing more methods. In the end, I always create it and use only when needed. It's "free" if you don't have to use it. 
I do not need to know the sequence ID name because we use `"id"`everywhere :smile: